### ISSUE TITLE
Correct way of adding special tokens before training a tokenizer

### ISSUE THREAD

**Issue by marcmk6** [huggingface/tokenizers#985]
Hi,

I want to train a tokenizer with code like the following
```
# I am not sure about the correct way, so I try to add '<sep>' in every possible way.

trainer = BpeTrainer(special_tokens=["<unk>", "<pad>", '<sep>'], vocab_size=vocab_size, )
tokenizer.add_special_tokens(['<sep>'])
tokenizer.add_tokens(['<sep>'])
tokenizer.train_from_iterator(trainer=trainer, iterator=iterator_over_seqs)
```
An example sequence is `ABCD<sep>EFGH`. 
However the trained vocabulary contains token `'<'`, `'>'` and `'e'`, `'ep'`, `'p'`, `'s'`, `'sep'`, which are undesired.

So I'm wondering what should I do to let the tokenizer taking the `'<sep>'` as a single special token?

Thanks

**Comment by Narsil** [comment]
Hi @marcmk6 ,

But does the tokenizer use those "undesired" tokens ? when submitting `<sep>` it definitely shouldn't. But the fact that those tokens exist might be entirely natural since they are needed for other regular text like `<separator>` maybe ?

I tried to feed data that didn't contain `<` and indeed it's not sued, can you try ?

**Comment by marcmk6** [comment]
> Hi @marcmk6 ,
> 
> But does the tokenizer use those "undesired" tokens ? when submitting `<sep>` it definitely shouldn't. But the fact that those tokens exist might be entirely natural since they are needed for other regular text like `<separator>` maybe ?
> 
> I tried to feed data that didn't contain `<` and indeed it's not sued, can you try ?

Hi @Narsil,

Thanks for your reply.
It seems the `<sep>` token doesn't get tokenized when using the trained tokenizer.

But what's confusing is that the only special token in the text that the tokenizer is trained on is `<sep>`. (I constructed these texts by myself so this is for sure.) Despite `<sep>`, the vocabulary can be as simple as the 26 capital letters `A` to `Z`. 
So I'm wondering why the `<sep>` is somewhat "tokenized" and these parts are included into the vocabulary, which is an unexpected behavior and it's misleading. 

Thanks again

**Comment by Narsil** [comment]
Can you try this:

```python
from tokenizers import Tokenizer, trainers, models, BertWordPieceTokenizer, pre_tokenizers

vocab_size = 8000
tokenizer = Tokenizer(models.BPE())
tokenizer.pre_tokenizer = pre_tokenizers.WhitespaceSplit()
trainer = trainers.BpeTrainer(
    special_tokens=["<unk>", "<pad>", "<sep>"],
    unk_token="<unk>",
    vocab_size=vocab_size,
)
tokenizer.add_special_tokens(["<sep>"])
tokenizer.add_tokens(["<sep>"])


def iterator():
    for i in range(1000):
        yield "This is a test"


tokenizer.train_from_iterator(iterator(), trainer=trainer)

tokenizer.save("tok.json", pretty=True)
tokenizer = Tokenizer.from_file("tok.json")

print(tokenizer.encode("test <sep> This").tokens)
```

And/or update it to show how we can make it fail ? (This was run using `0.12.1`)

**Comment by marcmk6** [comment]
```python
from tokenizers import Tokenizer, trainers, models, BertWordPieceTokenizer, pre_tokenizers

vocab_size = 8000
tokenizer = Tokenizer(models.BPE())
tokenizer.pre_tokenizer = pre_tokenizers.WhitespaceSplit()
trainer = trainers.BpeTrainer(
    special_tokens=["<unk>", "<pad>", "<sep>"],
    unk_token="<unk>",
    vocab_size=vocab_size,
)
tokenizer.add_special_tokens(["<sep>"])
tokenizer.add_tokens(["<sep>"])


def iterator():
    for i in range(1000):
        # yield "This is a test"
        yield "This <sep> is a test"


tokenizer.train_from_iterator(iterator(), trainer=trainer)

tokenizer.save("tok.json", pretty=True)
tokenizer = Tokenizer.from_file("tok.json")

print(tokenizer.encode("test <sep> This").tokens)
print(sorted(tokenizer.get_vocab().keys()))
```
Let's add a `<sep>` token in the training text and I got
```python
>>> print(tokenizer.encode("test <sep> This").tokens)
['test', '<sep>', 'This']
>>> print(sorted(tokenizer.get_vocab().keys()))
['<', '<pad>', '<s', '<sep', '<sep>', '<unk>', '>', 'T', 'Th', 'This', 'a', 'e', 'ep', 'es', 'h', 'i', 'is', 'p', 's', 't', 'tes', 'test']
```
The tokenization behaves as expected, as in my use case.
But you can see the vocabulary contains tokens like `<s` and `<sep`, which will never appear alone in texts like the training text.
(This was run on `0.11.6`.)

**Comment by Narsil** [comment]
Thanks for reproducing, also happens on  `master`.

This is because the `trainers` are fed the raw sentences, and NOT something that has been pre splitted with the `added tokens`.

Fixing this will require a big breaking change in the training unfortunately, so it will take time at the very least.

**Comment by marcmk6** [comment]
Ok I see. I just want to make sure this behavior is under expectation. Thanks for your help.

**Comment by Narsil** [comment]
I still think this is a bug, but one that will take a big amount of care to fix (if we ever do), not because of a technical challenge, but just because of the repercussions.

I even expected it to work like you said, so it's definitely something that could be modified.

**Comment by david-waterworth** [comment]
It would be nice if this could be fixed, I wanted to use `Replace` to replace certain strings with special tokens (i.e. \d+ with <digits>) but because the model treats these as characters rather than a single token it doesn't work reliably.

**Comment by Narsil** [comment]
@david-waterworth .
`Replace` the normalizer ? If yes, I don't think it should work.
`Addedtokens/AddedVocabulary` should happen before anything else (well it doesn't happen when feeding the trainer, but  it does during inference, and it probably should for the trainer too, but as I mentionned ,it's going to be tricky to handle the consequences maybe.)

`Replace` is able to handle regex, but this is not the issue right ?
Do you mind opening an issue with a smallish example of what you intend to do ?
